swarm/network/stream/testing/testing.go

```diff
@@ -0,0 +1,268 @@
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package testing
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"math/rand"
+	"os"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/p2p/discover"
+	"github.com/ethereum/go-ethereum/p2p/simulations"
+	"github.com/ethereum/go-ethereum/p2p/simulations/adapters"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/ethereum/go-ethereum/swarm/network"
+	"github.com/ethereum/go-ethereum/swarm/storage"
+)
+
+type Simulation struct {
+	Net    *simulations.Network
+	Stores []storage.ChunkStore
+	Addrs  []network.Addr
+	IDs    []discover.NodeID
+}
+
+func SetStores(addrs ...network.Addr) ([]storage.ChunkStore, func(), error) {
+	var datadirs []string
+	stores := make([]storage.ChunkStore, len(addrs))
+	var err error
+	for i, addr := range addrs {
+		var datadir string
+		datadir, err = ioutil.TempDir("", "streamer")
+		if err != nil {
+			break
+		}
+		var store storage.ChunkStore
+		store, err = storage.NewTestLocalStoreForAddr(datadir, addr.Over())
+		if err != nil {
+			break
+		}
+		datadirs = append(datadirs, datadir)
+		stores[i] = store
+	}
+	teardown := func() {
+		for i, datadir := range datadirs {
+			stores[i].Close()
+			os.RemoveAll(datadir)
+		}
+	}
+	return stores, teardown, err
+}
+
+func NewAdapter(adapterType string, services adapters.Services) (adapter adapters.NodeAdapter, teardown func(), err error) {
+	teardown = func() {}
+	switch adapterType {
+	case "sim":
+		adapter = adapters.NewSimAdapter(services)
+	case "socket":
+		adapter = adapters.NewSocketAdapter(services)
+	case "exec":
+		baseDir, err0 := ioutil.TempDir("", "swarm-test")
+		if err0 != nil {
+			return nil, teardown, err0
+		}
+		teardown = func() { os.RemoveAll(baseDir) }
+		adapter = adapters.NewExecAdapter(baseDir)
+	case "docker":
+		adapter, err = adapters.NewDockerAdapter()
+		if err != nil {
+			return nil, teardown, err
+		}
+	default:
+		return nil, teardown, errors.New("adapter needs to be one of sim, socket, exec, docker")
+	}
+	return adapter, teardown, nil
+}
+
+func CheckResult(t *testing.T, result *simulations.StepResult, startedAt, finishedAt time.Time) {
+	t.Logf("Simulation passed in %s", result.FinishedAt.Sub(result.StartedAt))
+	if len(result.Passes) > 1 {
+		var min, max time.Duration
+		var sum int
+		for _, pass := range result.Passes {
+			duration := pass.Sub(result.StartedAt)
+			if sum == 0 || duration < min {
+				min = duration
+			}
+			if duration > max {
+				max = duration
+			}
+			sum += int(duration.Nanoseconds())
+		}
+		t.Logf("Min: %s, Max: %s, Average: %s", min, max, time.Duration(sum/len(result.Passes))*time.Nanosecond)
+	}
+	t.Logf("Setup: %s, Shutdown: %s", result.StartedAt.Sub(startedAt), finishedAt.Sub(result.FinishedAt))
+}
+
+type RunConfig struct {
```

Comment:
should just be Config and should contain if nodes should be connected 
in a chain, spokes, ring, or custom snapshot